obstacledetection/src/main/java/com/caneseeproject/obstacledetection/ObstacleDetectorImpl.kt

```diff
@@ -1,63 +1,81 @@
 package com.caneseeproject.obstacledetection
 
+import com.caneseeproject.sensorPortals.PortalTranslator
 import com.caneseeproject.sensorPortals.Sensor
 import com.caneseeproject.sensorPortals.SensorPortal
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.filterIsInstance
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.map
 
 
-internal class ObstacleDetectorImpl(private val odPortal: SensorPortal) : ObstacleDetector {
-
-    private lateinit var cane: Sensor
+internal class ObstacleDetectorImpl(private val odPortal: SensorPortal,
+                                    private val odTranslator: PortalTranslator<ODReading, ODControl> = ODTranslator() )
+    : ObstacleDetector {
 
     override fun activate() {
-        cane = odPortal.connect()
+      odPortal.open()
+    }
+
+    override suspend fun send(vararg messages: ODControl) {
+        odPortal.send(*(messages.map { odTranslator.pack(it) }.toTypedArray()))
+    }
+
+    override fun readings(): Flow<ODReading> =
+        odPortal.receive().map { odTranslator.tokenize(it) }.filterIsInstance()
```

Comment:
since `readings` returns `ODReading`, no need to `filterIsInstance`, but you need to `filterNotNull`.

Reply:
hmmmmmm, if it is an instance , then it ain't null, right ??

Reply:
wait , no!